### ISSUE TITLE
HttpClient and Activity lifetime

### ISSUE THREAD

**Issue by JamesNK** [dotnet/runtime#37141]
It appears that HttpClient's HTTP request activity ends when HTTP headers are received. 

Discussion about it in OpenTelemetry - https://github.com/open-telemetry/opentelemetry-dotnet/pull/687#discussion_r429501398

I don't know whether this is a bug or intended behavior. Wouldn't it be more accurate for the activity to end with the response finishing?

**Comment by davidsh** [comment]
>I don't know whether this is a bug or intended behavior. Wouldn't it be more accurate for the activity to end with the response finishing?

In general, it is by design. Many of the HttpClient APIs will return immediately after reading just the headers, i.e.  [HttpCompletionOption.ResponseHeadersRead](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpcompletionoption?view=netcore-3.1).

Also, the design of the HttpClient and the handler pipeline based on HttpMessageHandler is that the pipeline work by getting an HttpRequestMessage and returning an HttpResponseMessage. And handlers returns an HttpResponseMessage without reading the response body.

**Comment by JamesNK** [comment]
Ok. It seems inconsistent with how requests in other network graphs are displayed, but I don't have a requirement to change it. The gRPC client has its own activity that does run until the response is complete.

**Comment by davidsh** [comment]
Re-activating as per https://github.com/dotnet/runtime/pull/37619#discussion_r439597451.

This issue here is primarily about how [DiagnosticHandler](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs) works (which is based on DiagnosticSource and not EventSource).

**Comment by karelz** [comment]
@davidsh are you working on this as part of YARP Telemetry?

**Comment by davidsh** [comment]
>@davidsh are you working on this as part of YARP Telemetry?

I am not currently.

DiagnosticHandler is separate from YARP Telemetry. YARP Telemetry is about adding new EventSource and EventCounters for telemetry. This issue is about fixing (changing the behavior) of the existing DiagnosticSource events so that RequestStop is issued when the request body is fully processed. That will be non-trivial for DiagnosticHandler because as a handler for HttpClient, it only knows about response headers. To fix this issue, DiagnosticHandler will need to wrap HttpResponseMessage.Content objects in some way to get access to the response body and trigger an event when the response body is fully processed. However, an alternate solution is possible... it could piggyback on my work in PR #37619 by adding an internal EventSource listener to the RequestStop event that will be done for telemetry.

**Comment by scalablecory** [comment]
@JamesNK how critical is this for you? .NET 5? or can we push it?

**Comment by JamesNK** [comment]
It's not important. Push away.

**Comment by JamesNK** [comment]
This is what HttpClient telemetry looks like today when captured:

![image](https://user-images.githubusercontent.com/303201/85946627-c08fb600-b999-11ea-8532-46eede031dc2.png)

It stops as soon as the headers are received, rather than waiting until the server has finished the response.

**Comment by ManickaP** [comment]
@antonfirsov this seems like it might be moot with all the recent work in line with OpenTelemetry sementics. Should we close this?

**Comment by antonfirsov** [comment]
Indeed, it is extremely unlikely we will ever change this. We may consider a new activity if there will be serious end-user feedback requesting it in new issue(s).
